### ISSUE TITLE
Main event silent failure

### ISSUE THREAD

**Issue by hopeyen** [graphops/poi-radio#115]
**Describe the bug**
From indexer Suntzu's report, we found that the radio can fail or get stuck silently. 

The most recent logs were 2 days before the time of report which the radio has not exited the program. The most recent 100 logs are all gowaku node, and it stopped when the last peer was disconnected.
```
output:
2023-04-25T16:59:09.106Z        INFO    gowaku.node2.filter     filter/waku_filter.go:137       received a message push {"fullNode": false, "peer": "16Uiu2HAm5uqfdh7z2YTEps2MhvsXTk3uvSHZ9AtVkzipZZGbKJEL", "messages": 1}
2023-04-25T16:59:16.307Z        INFO    gowaku.node2    node/connectedness.go:68        peer disconnected       {"peer": "16Uiu2HAm5uqfdh7z2YTEps2MhvsXTk3uvSHZ9AtVkzipZZGbKJEL"}
```

In theory, main loop will periodically update the states of network and subgraphs with logs printed, and will reconnect to the peers that went on/offline. We don't see that here so the event might be stuck/silently failed. 

**Expected behavior**
Radio should
- Fail or retry later if a operational process panics
- Timeout and retry later if a process is stuck

**Screenshots**
![Screenshot_2023-04-27_at_13 48 25](https://user-images.githubusercontent.com/60078528/234788768-e0da237c-d524-4ac6-9af0-1707da2a1c60.png)

**Comment by axiomatic-aardvark** [comment]
Sadly I haven't been able to reproduce this issue :( I tried the following:
- Switching the Radio's boot node configuration to use a single boot node running on a droplet, then stopping that boot node while the Radio is running to see what happens if it loses connection to all boot nodes
- Turning my internet off for 30s
- Leaving the instance running for 3h
- "Overloading" the network by running 50 instances at the same time with the same content topics

I also talked with @hopeyen and our current guess is that this might be caused by a deadlock issue. After all, we have a lot of concurrent `Mutex`es being accessed (both reading and writing) all over the place. Here are some actions we can take to reduce the risk of deadlocks, in theory:
- Where possible, switch to using `sleep` and `Mutex` from `tokio` rather than `std`, since the standard library ones aren't meant to async environments and increase the risk of issues.
- Add timeouts for all our `.await` calls

**Comment by axiomatic-aardvark** [comment]
Got a confirmation that **kw1k** from **Nodeify** that they also encountered this issue. On the other hand, **Vincent** from **Data Nexus** hasn't had any problems. So far everyone that has reported seeing this issue is running POI Radio as a Docker image, whereas users that are running the Radio bare metal (using a release binary) are not seeing it.

**Comment by axiomatic-aardvark** [comment]
kw1k suggested that when there are silent failures in Docker it could be related to resources issues. We are indeed clotting up too much in-memory data and maybe need to clear more often/more thoroughly

**Comment by axiomatic-aardvark** [comment]
stake-machine (also running POI Radio in Docker) reported seeing this issue on testnet, but not on mainnet

**Comment by hopeyen** [comment]
hey there @stake-machine ! We have recently released a new version [0.2.0](https://github.com/graphops/poi-radio/pkgs/container/poi-radio) that should solve the issue described here. 
There might be more deadlocks the program might run into, but a more focused refactoring of threads is coming up soon; it would be great if you could give 0.2.0 a try and help observe the behavior.

**Comment by hopeyen** [comment]
Closing for now, feel free to re-open or open a new issue if the radio meets a new deadlock 🔮
